cpp/src/arrow/array/util.cc

```diff
@@ -206,9 +209,9 @@ class ArrayDataEndianSwapper {
     auto data = reinterpret_cast<const MonthDayNanos*>(data_->buffers[1]->data());
     ARROW_ASSIGN_OR_RAISE(auto new_buffer, AllocateBuffer(data_->buffers[1]->size()));
     auto new_data = reinterpret_cast<MonthDayNanos*>(new_buffer->mutable_data());
-    int64_t length = data_->length;
+    const int64_t length = data_->buffers[1]->size() / sizeof(MonthDayNanos);
```

Comment:
it might be worth adding a comment (here and elsewhere) similar to what is tested (at this point the data isn't trusted yet so we can't rely on metadata?